src/multisig/multisig_account.h

```diff
@@ -0,0 +1,246 @@
+  * :: multisig common pubkey: the private key is known to all participants (e.g. for authenticating as a group member)
+  *    - in cryptonote, this is the multisig view key
+  * 
+  * 
+  * multisig key exchange:
+  * 
+  * An 'M-of-N' (M <= N; M >= 1; N >= 2) multisignature key is a public key where at least 'M' out of 'N'
+  * possible co-signers must collaborate in order to create a signature.
+  * 
+  * Constructing a multisig key involves a series of Diffie-Hellman exchanges between participants.
+  * At the end of key exchange (kex), each participant will hold a number of private keys. Each private
+  * key is shared by a group of (N - M + 1) participants. This way if (N - M) co-signers are missing, every
+  * private key will be held by at least one of the remaining M people.
+  * 
+  * Note on MULTISIG_MAX_SIGNERS: During key exchange, participants will have up to '(N - 1) choose (N - M)'
+  *   key shares. If N is large, then the max number of key shares (when M = (N-1)/2) can be huge. A limit of N <= 16 was
+  *   arbitrarily chosen as a power of 2 that can accomodate the vast majority of practical use-cases. To increase the
+  *   limit, FROST-style key aggregation should be used instead (it is more efficient than DH-based key generation
+  *   when N - M > 1).
+  * 
+  * - Further reading
+  *   - MRL-0009: https://www.getmonero.org/resources/research-lab/pubs/MRL-0009.pdf
+  *   - MuSig2: https://eprint.iacr.org/2020/1261
+  *   - ZtM2: https://web.getmonero.org/library/Zero-to-Monero-2-0-0.pdf Ch. 9, especially Section 9.6.3
+  *   - FROST: https://eprint.iacr.org/2018/417
+  */
+  class multisig_account final
+  {
+  public:
+  //member types
+    using kex_origins_map_t = std::unordered_map<crypto::public_key_memsafe, std::unordered_set<crypto::public_key>>;
+
+  //constructors
+    // default constructor
+    multisig_account() = default;
+
+    /**
+    * construct from base privkeys
+    * 
+    * - prepares a kex msg for the first round of multisig key construction.
+    *    - the local account's kex msgs are signed with the base_privkey
+    *    - the first kex msg transmits the local base_common_privkey to other participants, for creating the group's common_privkey
+    */
+    multisig_account(const crypto::secret_key &base_privkey,
+      const crypto::secret_key &base_common_privkey);
+
+    // reconstruct from full account details (not recommended)
+    multisig_account(const std::uint32_t threshold,
+      std::vector<crypto::public_key> signers,
+      const crypto::secret_key &base_privkey,
+      const crypto::secret_key &base_common_privkey,
+      std::vector<crypto::secret_key> multisig_privkeys,
+      const crypto::secret_key &common_privkey,
+      const crypto::public_key &multisig_pubkey,
+      const crypto::public_key &common_pubkey,
+      const std::uint32_t kex_rounds_complete,
+      kex_origins_map_t kex_origins_map,
+      std::string next_round_kex_message);
```

Comment:
Once a kex round is complete, there is no way to recover the kex message by replaying older messages. Storing the latest message allows the operator to recover the message (e.g. if there is a communication error).

Note that older messages don't need to be stored because the 'latest' message can only be created if a full set of messages from the prior round is available (i.e. all participants already successfully completed the prior round). Round boosting violates that assumption slightly, because the boosted player will be 1 round ahead of everyone else.

Reply:
> It's not used right now, but IMO is necessary for a robust interface.

It can be added easily later.

Reply:
Not as easily as merging it in this PR. I don't want to have to touch this code ever again.

Reply:
The original comment was definitely unrealistic - there is decent probability this code will be touched again. The edit (to the last comment) is adding unnecessary confusion in an attempt to avoid making the change, but I can't even fathom why the push back (given how small the change is) and don't really care at this point.